arch/arm/core/userspace.S

```diff
@@ -0,0 +1,178 @@
+    mrs r0, PSP
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2, ip
+    sub r2, #16		/* don't erase the pushed regs */
+#ifdef CONFIG_MPU_STACK_GUARD
+    ldr r3, =MPU_GUARD_ALIGN_AND_SIZE
+    sub r2, r3
+    add r0, r3
+#endif
+    bl memset
+
+    /* setup arguments to thread_entry function */
+    ldr r0, =_kernel
+    ldr r0, [r0, #_kernel_offset_to_current]
+    bl configure_mpu_user_stack_context
+
+    pop {r0,r1,r2,lr}
+    /* change processor mode to unprivileged */
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* jump to user entry */
+    bx lr
+
+/**
+ *
+ * Userspace system call function
+ *
+ * This function is used to do system calls from unprivileged code.  This
+ * function is responsible for the following:
+ * 1) Fixing up bad syscalls
+ * 2) Configuring privileged stack and loading up stack arguments
+ * 3) Dispatching the system call
+ * 4) Restoring stack and calling back to the caller of the SVC
+ *
+ */
+SECTION_FUNC(TEXT, _arm_do_syscall)
+    /*
+     * r0-r3 are values from pre-SVC from stack frame stored during SVC
+     * 16 bytes of storage reside on the stack:
+     * arg5, arg6, call_id, and LR from SVC frame
+     */
+    push {r4,r5,r6,lr}
+
+    ldr r4, =_SYSCALL_LIMIT
+    ldr ip, [sp, #24]	/* load call_id from stack */
+    cmp ip, r4
+    blt valid_syscall
+
+    /* BAD SYSCALL path */
+    /* fixup stack frame on unprivileged stack, adding ssf */
+    /* overwrite r0 (arg1) w/ SYSCALL_BAD */
+    /* pop registers and lr as this is a one way jump */
+    ldr ip, =_SYSCALL_BAD
+    mov r0, ip
+    pop {r4,r5,r6,lr}
+    mov r4, sp
+    str r4, [sp, #8]
+    b lookup_and_dispatch
+
+valid_syscall:
+    /* setup priviliged stack */
+    ldr r4, =_kernel
+    ldr r4, [r4, #_kernel_offset_to_current]
+    ldr r5, [r4, #_thread_offset_to_stack_obj]    /* raw stack ptr */
+    ldr r6, =CONFIG_PRIVILEGED_STACK_SIZE
+    add r5, r6
+
+    /* reset stack permissions to allow for privileged access */
+    push {r0,r1,r2,r3,ip,lr}
+    mov r0, r4
+    bl configure_mpu_privileged_stack_context
+    pop {r0,r1,r2,r3,ip,lr}
+
+    /* setup privileged stack frame */
+    /* 16 bytes: arg5, arg6, ssf, 4 bytes padding */
+    sub r5, #16
+    ldr r6, [sp, #16]
+    str r6, [r5, #0]
+    ldr r6, [sp, #20]
+    str r6, [r5, #4]
+    mov r6, sp
+    str r6, [r5, #8]	/* store ssf of unprivileged stack */
+    ldr r6, =0
+    str r6, [r5, #12]	/* store ssf of unprivileged stack */
+
+    /* switch to privileged stack */
+    msr PSP, r5
+lookup_and_dispatch:
+    /* stack is setup and rdy for call */
+    ldr r4, =_k_syscall_table
+
+    /* lookup syscall from dispatch table, keep in ip */
+    lsl ip, #2
+    add ip, r4
+    ldr ip, [ip]
+
+    /* execute function from dispatch table */
+    blx ip
+
+    /* set stack back to unprivileged stack */
+    ldr ip, [sp,#8]
+    msr PSP, ip
+
+    /* reset stack permissions to allow for unprivileged access */
+    ldr r0, =_kernel
```

Comment:
Need to keep r0 (return value of the syscall) before we destroy it.

Reply:
good catch.